lib/cli/test/unit/Cardano/CLISpec.hs

```diff
@@ -473,10 +558,132 @@ spec = do
+            propCliKeySchemeEquality
+                (byronCliKeyScheme s)
+                (byronCliKeyScheme s)
+
+        -- This tests provides a stronger guarantee than merely knowing that
+        -- unsafeHexTextToXPrv and xPrvToHexText roundtrips.
+        it "scheme == mapKey (fromHex . toHex) scheme"
+            $ property prop_roundtripCliKeySchemeKeyViaHex
+
+        it "random /= icarus" $ do
+            expectFailure $ propCliKeySchemeEquality
+                (byronCliKeyScheme Random)
+                (byronCliKeyScheme Icarus)
+
   where
     backspace :: Text
     backspace = T.singleton (toEnum 127)
 
+prop_roundtripCliKeySchemeKeyViaHex :: ByronWalletStyle -> Property
+prop_roundtripCliKeySchemeKeyViaHex style =
+            propCliKeySchemeEquality
+                (byronCliKeyScheme style)
+                (mapKey unsafeHexTextToXPrv
+                    . mapKey xPrvToHexText
+                    $ byronCliKeyScheme style)
+
+prop_allowedWordLengthsAllWork :: ByronWalletStyle -> Property
+prop_allowedWordLengthsAllWork style = do
+    (forAll (genAllowedMnemonic s) propCanRetrieveRootKey)
+  where
+    s :: CliKeyScheme XPrv (Either String)
+    s = byronCliKeyScheme style
+
+    propCanRetrieveRootKey :: [Text] -> Property
+    propCanRetrieveRootKey mw = case mnemonicToRootKey s mw of
+        Right _ -> property True
+        Left e -> counterexample
+            (show (length mw) ++ " words, failed with: " ++ e)
+            (property False)
+
+propCliKeySchemeEquality
+    :: CliKeyScheme XPrv (Either String)
+    -> CliKeyScheme XPrv (Either String)
+    -> Property
+propCliKeySchemeEquality s1 s2 = do
+    (forAll (genAllowedMnemonic s1) propSameMnem)
+    .&&.
+    (allowedWordLengths s1) === (allowedWordLengths s2)
+  where
+    propSameMnem :: [Text] -> Property
+    propSameMnem mw = (mnemonicToRootKey s1 mw) === (mnemonicToRootKey s2 mw)
+
+genAllowedMnemonic :: CliKeyScheme key m -> Gen [Text]
+genAllowedMnemonic s = oneof (map genMnemonicOfSize $ allowedWordLengths s)
+
+genMnemonicOfSize :: Int -> Gen [Text]
+genMnemonicOfSize 12 = mnemonicToText <$> genMnemonic @12
+genMnemonicOfSize 15 = mnemonicToText <$> genMnemonic @15
+genMnemonicOfSize 18 = mnemonicToText <$> genMnemonic @18
+genMnemonicOfSize 21 = mnemonicToText <$> genMnemonic @21
+genMnemonicOfSize 24 = mnemonicToText <$> genMnemonic @24
+genMnemonicOfSize n = error $ "when this test was written, "
+    ++ show n ++ " was not a valid length of a mnemonic"
+
+instance Show XPrv where
+    show = show . unXPrv
+
+instance Eq XPrv where
+    a == b = unXPrv a == unXPrv b
+
+genMnemonic
+    :: forall mw ent csz.
+     ( ConsistentEntropy ent mw csz
+     , EntropySize mw ~ ent
+     )
+    => Gen (Mnemonic mw)
+genMnemonic = do
+        let n = fromIntegral (natVal $ Proxy @(EntropySize mw)) `div` 8
+        bytes <- BS.pack <$> vectorOf n arbitrary
+        let ent = either (error . show) id $ mkEntropy @(EntropySize mw) bytes
```

Comment:
what about `unsafeMkEntropy` ?

Reply:
Indeed 👍 

That said, I'm starting to question how valuable the `unsafe...` family of functions really are. `mkEntropy`,  `EntropySize`, etc.  come from `Cardano.Wallet.Primitive.Mnemonic`, whereas `unsafeMkEntropy` now is the odd one coming from `Cardano.Wallet.Unsafe`.

I guess they make more sense when they're able to `f . g . either (error . show) id .  h`, rather than just `either (error . show) id . f` like in this specific example.

… And I think there's _something_ attractive about having a multitude of various `unsafe...` functions all in the _same_ module. Maybe it's easier to import `unsafeFoo` from `Cardano.Wallet.Unsafe` than remember where the original `foo`-function is defined  when you're writing a test.

Reply:
Note that the `unsafe` function are only convenience. We (ought to) only use them in test code for either literal values or values that are hand-crafted and known to be correct. The real question is "why is this module part of the library and not the test-suite only". And the main reason for this is just that it's shared across multiple test modules (for what the argument is worth).

Reply:
> Note that the unsafe function are only convenience.

Well, I didn't really see the convenience in writing `unsafeFoo` rather than `either (error . show) id foo`, or `unsafe foo`.